fuse-ext2.1
.TH FUSE-EXT2 "1" "September 2009" "FUSE/UMFUSE modules" "User Commands"
.SH NAME
FUSE-EXT2 \- FUSE module for Second Extended File System
.SH SYNOPSIS
.B fuse-ext2
.RI [ OPTION ]...\&
.I imagefile mountpoint
.LP
.B fuse-ext2
.I imagefile mountpoint
.RI [ OPTION ]...
.SH DESCRIPTION
Fuse-ext2 mounts an ext2/ext3/ext4 partition or image file. It uses Fuse support
(Filesystem in Userspace) included in modern Linux kernels.
The same module is available for view-os (umview(1), kmview(1)) under
the name \fBumfuseext2\fR.
.SH OPTIONS
.SS "General options"
.TP
\fB\-o\fR opt,[opt...]
mount options
.TP
\fB\-h\fR   \fB\-\-help\fR
print help
.TP
\fB\-V\fR   \fB\-\-version\fR
print version
.SS "FUSE-EXT2 options:"
.TP
\fB\-o\fR force
force read-write mount (\fBEXPERIMENTAL\fR, the option \fBrw\fR without \fBforce\fR
mounts the file system in read-only mode, with a warning).
.TP
\fB\-o\fR rw+
enable read-write mount (\fBEXPERIMENTAL\fR, is a shortcut for -o rw,force)
.SS "FUSE options:"

.TP
\fB\-d, \-o debug\fR
enable debug output (implies \fB\-f\fR)
.TP
\fB\-f\fR
foreground operation
.TP
\fB\-s\fR
disable multi\-threaded operation
.TP
\fB\-o allow_other\fR
allow access to other users
.TP
\fB\-o allow_root\fR
allow access to root
.TP
\fB\-o nonempty\fR
allow mounts over non\-empty file/dir
.TP
\fB\-o default_permissions\fR
enable permission checking by kernel
.TP
\fB\-o fsname=\fINAME\fR
set filesystem name
.TP
\fB\-o large_read\fR
issue large read requests (2.4 only)
.TP
\fB\-o max_read=\fIN\fR
set maximum size of read requests
.TP
\fB\-o hard_remove\fR
immediate removal (don't hide files)
.TP
\fB\-o use_ino\fR
let filesystem set inode numbers
.TP
\fB\-o readdir_ino\fR
try to fill in d_ino in readdir
.TP
\fB\-o direct_io\fR
use direct I/O
.TP
\fB\-o kernel_cache\fR
cache files in kernel
.TP
\fB\-o [no]auto_cache\fR
enable caching based on modification times
.TP
\fB\-o umask=\fIM\fR
set file permissions (octal)
.TP
\fB\-o uid=\fIN\fR
set file owner
.TP
\fB\-o gid=\fIN\fR
set file group
.TP
\fB\-o entry_timeout=\fIT\fR
cache timeout for names (1.0s)
.TP
\fB\-o negative_timeout=\fIT\fR
cache timeout for deleted names (0.0s)
.TP
\fB\-o attr_timeout=\fIT\fR
cache timeout for attributes (1.0s)
.TP
\fB\-o ac_attr_timeout=\fIT\fR
auto cache timeout for attributes (attr_timeout)
.TP
\fB\-o intr\fR
allow requests to be interrupted
.TP
\fB\-o intr_signal=\fINUM\fR
signal to send on interrupt (10)
.TP
\fB\-o max_write=\fIN\fR
set maximum size of write requests
.TP
\fB\-o max_readahead=\fIN\fR
set maximum readahead
.TP
\fB\-o async_read\fR
perform reads asynchronously (default)
.TP
\fB\-o sync_read\fR
perform reads synchronously
.SH AUTHORS
.TP
Alper Akcan <alper.akcan@gmail.com>
.TP
Renzo Davoli <renzo@cs.unibo.it>
.TP
See \fIhttp://github.com/alperkacan/fuse-ext2/, http://www.virtualsquare.org\fR.


ChangeLog
Mon Mar 30 19:17:44 EEST 2015
	- version 0.0.9
	- fix write tests for pjdfstest
	- fix ext2fs_file_*, need to open/release in every file operation

Mon Mar 28 10:21:43 EEST 2015
	- version 0.0.8 - lazarus, raised from dead
	- use e2fsprogs from system (mac, linux)
	- fix build routines for new systems (mac, linux)
	- switch to osxfuse (mac)

Wed 30 Dec 2009 04:17:02 EET
	- enable noappledouble only for readonly mode by default

Thu 24 Dec 2009 14:40:10 EET
	- version 0.0.7

Tue 03 Nov 2009 04:02:06 AM EEST
	- get device volume name in probe stage. now diskutil shows volume name
	  on device bar, as a result device mount point is created as
	  '/Volumes/${volume_name}'.
	- supports mac osx 10.6 in compatibility mode.

Fri 09 Oct 2009 12:53:02 AM EEST
	- fixed big file support (write > 4G)

Tue 29 Sep 2009 05:25:08 AM EEST
	- improved probe time (mount time improved %50)
	- improved mount time for read-only mode
	- mount time reduced from ~8 secs to ~2 secs
	- added windows_fat_(16/32) partition type to probe types
	- added 0.0.6 md5sum to release.xml
	- added release-beta.xml, just like macfuse
	- linux fs type probe decreased from 50 to 3500
	- disabled apple double files

Wed 23 Sep 2009 00:15:10 AM EEST
	- version 0.0.6
	- fixed big file support for size > 4Gb
	- prefpane supports uninstall/update

Mon 07 Sep 2009 00:02:40 AM EEST
	- added fuse-ext2.1 man page written by Renzo

Wed 02 Sep 2009 07:43:48 PM EEST
	- merge in renzo branch rev.140 including:
	- support of file truncate
	- support of concurrent file access
	- view-os compatibility (see wiki.virtualsquare.org, umfuse)

Mon 22 Jun 2009 05:36:45 PM EEST
	- uninstall support for mac
	- initial updater for mac

Thu 11 Jun 2009 10:55:23 PM EEST
	- set version information at compile time for info.plist
	- corrected mke2fs path information in make-pkg.sh

Tue 09 Jun 2009 02:09:15 AM EEST
	- upgrade to e2fsprogs-1.41.6

Mon 08 Jun 2009 12:47:59 AM EEST
	- added uuid lib

Wed 3 Jun 2009 07:09:27 PM EEST
	- version 0.0.5
	- updated authors file
	- minor directory changes

Tue 2 Jun 2009 05:04:49 PM EEST
	- big/little endian support for universal build

Tue 2 Jun 2009 00:39:06 AM EEST
	- upgrade to e2fsprogs 1.41.5
	- added mke2fs for disk format
	- added format support to diskutil

Wed 27 May 2009 03:11:38 AM EEST
	- 10.4 tiger build instructions added to README
	- added /usr/local/[lib,include] to fuse-ext2
	  compile flags.

Tue 19 May 2009 07:34:23 PM EEST
	- version 0.0.4
	- get etx2 label in probe stage
	- make labeling work for universal build, also

Tue 21 Apr 2009 02:32:57 PM EEST
	- merge in Dave Vasilevsky's labels patch

Sat 04 Apr 2009 02:28:34 AM EET
	- disable multithread
	- reduce wait time to 5 seconds

Mon 09 Feb 2009 04:42:51 PM EET
	- mac osx file system bundle support
	- automount support
	- first 'real' stable (read only)


fuse-ext2/fuse-ext2.c
/**
 * Copyright (c) 2008-2015 Alper Akcan <alper.akcan@gmail.com>
 * Copyright (c) 2009 Renzo Davoli <renzo@cs.unibo.it>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (in the main directory of the fuse-ext2
 * distribution in the file COPYING); if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "fuse-ext2.h"

static const char *HOME = "http://github.com/alperakcan/fuse-ext2/";

#if __FreeBSD__ == 10
static char def_opts[] = "allow_other,default_permissions,local,";
static char def_opts_rd[] = "noappledouble,";
#else
static char def_opts[] = "allow_other,default_permissions,";
static char def_opts_rd[] = "";
#endif

static const char *usage_msg =
"\n"
"%s %s %d - FUSE EXT2FS Driver\n"
"\n"
"Copyright (C) 2008-2015 Alper Akcan <alper.akcan@gmail.com>\n"
"Copyright (C) 2009 Renzo Davoli <renzo@cs.unibo.it>\n"
"\n"
"Usage:    %s <device|image_file> <mount_point> [-o option[,...]]\n"
"\n"
"Options:  ro, force, allow_others\n"
"          Please see details in the manual.\n"
"\n"
"Example:  fuse-ext2 /dev/sda1 /mnt/sda1\n"
"\n"
"%s\n"
"\n";

static int strappend (char **dest, const char *append)
{
	char *p;
	size_t size;

	if (!dest) {
		return -1;
	}
	if (!append) {
		return 0;
	}

	size = strlen(append) + 1;
	if (*dest) {
		size += strlen(*dest);
	}

	p = realloc(*dest, size);
    	if (!p) {
    		debugf_main("Memory realloction failed");
		return -1;
	}

	if (*dest) {
		strcat(p, append);
	} else {
		strcpy(p, append);
	}
	*dest = p;

	return 0;
}

static void usage (void)
{
	printf(usage_msg, PACKAGE, VERSION, fuse_version(), PACKAGE, HOME);
}

static int parse_options (int argc, char *argv[], struct extfs_data *opts)
{
	int c;

	static const char *sopt = "o:hv";
	static const struct option lopt[] = {
		{ "options",	 required_argument,	NULL, 'o' },
		{ "help",	 no_argument,		NULL, 'h' },
		{ "verbose",	 no_argument,		NULL, 'v' },
		{ NULL,		 0,			NULL,  0  }
	};

#if 0
	printf("arguments;\n");
	for (c = 0; c < argc; c++) {
		printf("%d: %s\n", c, argv[c]);
	}
	printf("done\n");
#endif

	opterr = 0; /* We'll handle the errors, thank you. */

	while ((c = getopt_long(argc, argv, sopt, lopt, NULL)) != -1) {
		switch (c) {
			case 'o':
				if (opts->options)
					if (strappend(&opts->options, ","))
						return -1;
				if (strappend(&opts->options, optarg))
					return -1;
				break;
			case 'h':
				usage();
				exit(9);
			case 'v':
				/*
				 * We must handle the 'verbose' option even if
				 * we don't use it because mount(8) passes it.
				 */
				opts->debug = 1;
				break;
			default:
				debugf_main("Unknown option '%s'", argv[optind - 1]);
				return -1;
		}
	}

	if (optind < argc) {
		optarg=argv[optind++];
		if (optarg[0] != '/') {
			char fulldevice[PATH_MAX+1];
			if (!realpath(optarg, fulldevice)) {
				debugf_main("Cannot mount %s", optarg);
				free(opts->device);
				opts->device = NULL;
				return -1;
			} else
				opts->device = strdup(fulldevice);
		} else
			opts->device = strdup(optarg);
	}

	if (optind < argc) {
		opts->mnt_point = argv[optind++];
	}

	if (optind < argc) {
		debugf_main("You must specify exactly one device and exactly one mount point");
		return -1;
	}

	if (!opts->device) {
		debugf_main("No device is specified");
		return -1;
	}
	if (!opts->mnt_point) {
		debugf_main("No mountpoint is specified");
		return -1;
	}

	return 0;
}

static char * parse_mount_options (const char *orig_opts, struct extfs_data *opts)
{
	char *options, *s, *opt, *val, *ret;

	ret = malloc(strlen(def_opts) + strlen(def_opts_rd) + strlen(orig_opts) + 256 + PATH_MAX);
	if (!ret) {
		return NULL;
	}

	*ret = 0;
	options = strdup(orig_opts);
	if (!options) {
		debugf_main("strdup failed");
		return NULL;
	}

	s = options;
	while (s && *s && (val = strsep(&s, ","))) {
		opt = strsep(&val, "=");
		if (!strcmp(opt, "ro")) { /* Read-only mount. */
			if (val) {
				debugf_main("'ro' option should not have value");
				goto err_exit;
			}
			opts->readonly = 1;
			strcat(ret, "ro,");
		} else if (!strcmp(opt, "rw")) { /* Read-write mount */
			if (val) {
				debugf_main("'rw' option should not have value");
				goto err_exit;
			}
			opts->readonly = 0;
			strcat(ret, "rw,");
		} else if (!strcmp(opt, "rw+")) { /* Read-write mount */
			if (val) {
				debugf_main("'rw+' option should not have value");
				goto err_exit;
			}
			opts->readonly = 0;
			opts->force = 1;
			strcat(ret, "rw,");
		} else if (!strcmp(opt, "debug")) { /* enable debug */
			if (val) {
				debugf_main("'debug' option should not have value");
				goto err_exit;
			}
			opts->debug = 1;
			strcat(ret, "debug,");
		} else if (!strcmp(opt, "silent")) { /* keep silent */
			if (val) {
				debugf_main("'silent' option should not have value");
				goto err_exit;
			}
			opts->silent = 1;
		} else if (!strcmp(opt, "force")) { /* enable read/write */
			if (val) {
				debugf_main("'force option should no have value");
				goto err_exit;
			}
			opts->force = 1;
#if __FreeBSD__ == 10
			strcat(ret, "force,");
#endif
		} else { /* Probably FUSE option. */
			strcat(ret, opt);
			if (val) {
				strcat(ret, "=");
				strcat(ret, val);
			}
			strcat(ret, ",");
		}
	}

	if (opts->readonly == 0 && opts->force == 0) {
		fprintf(stderr, "Mounting %s Read-Only.\nUse \'force\' or \'rw+\' options to enable Read-Write mode\n",opts->device);
		opts->readonly = 1;
	}

	strcat(ret, def_opts);
	if (opts->readonly == 1) {
		strcat(ret, def_opts_rd);
		strcat(ret, "ro,");
	}
	strcat(ret, "fsname=");
	strcat(ret, opts->device);
#if __FreeBSD__ == 10
	strcat(ret, ",fstypename=");
	strcat(ret, "ext2");
	strcat(ret, ",volname=");
	if (opts->volname == NULL || opts->volname[0] == '\0') {
		s = strrchr(opts->device, '/');
		if (s != NULL) {
			strcat(ret, s + 1);
		} else {
			strcat(ret, opts->device);
		}
	} else {
		strcat(ret, opts->volname);
	}
#endif
exit:
	free(options);
	return ret;
err_exit:
	free(ret);
	ret = NULL;
	goto exit;
}

static const struct fuse_operations ext2fs_ops = {
	.getattr        = op_getattr,
	.readlink       = op_readlink,
	.mknod          = op_mknod,
	.mkdir          = op_mkdir,
	.unlink         = op_unlink,
	.rmdir          = op_rmdir,
	.symlink        = op_symlink,
	.rename         = op_rename,
	.link           = op_link,
	.chmod          = op_chmod,
	.chown          = op_chown,
	.truncate       = op_truncate,
	.open           = op_open,
	.read           = op_read,
	.write          = op_write,
	.statfs         = op_statfs,
	.flush          = op_flush,
	.release	= op_release,
	.fsync          = op_fsync,
	.setxattr       = NULL,
	.getxattr       = NULL,
	.listxattr      = NULL,
	.removexattr    = NULL,
	.opendir        = op_open,
	.readdir        = op_readdir,
	.releasedir     = op_release,
	.fsyncdir       = op_fsync,
	.init		= op_init,
	.destroy	= op_destroy,
	.access         = op_access,
	.create         = op_create,
	.ftruncate      = op_ftruncate,
	.fgetattr       = op_fgetattr,
	.lock           = NULL,
	.utimens        = op_utimens,
	.bmap           = NULL,
};

int main (int argc, char *argv[])
{
	int err = 0;
	struct stat sbuf;
	char *parsed_options = NULL;
	struct fuse_args fargs = FUSE_ARGS_INIT(0, NULL);
	struct extfs_data opts;

	debugf_main("version:'%s', fuse_version:'%d / %d / %d'", VERSION, FUSE_USE_VERSION,  FUSE_VERSION, fuse_version());

	memset(&opts, 0, sizeof(opts));

	if (parse_options(argc, argv, &opts)) {
		usage();
		return -1;
	}

	if (stat(opts.device, &sbuf)) {
		debugf_main("Failed to access '%s'", opts.device);
		err = -3;
		goto err_out;
	}

	if (do_probe(&opts) != 0) {
		debugf_main("Probe failed");
		err = -4;
		goto err_out;
	}

	parsed_options = parse_mount_options(opts.options ? opts.options : "", &opts);
	if (!parsed_options) {
		err = -2;
		goto err_out;
	}

	debugf_main("opts.device: %s", opts.device);
	debugf_main("opts.mnt_point: %s", opts.mnt_point);
	debugf_main("opts.volname: %s", (opts.volname != NULL) ? opts.volname : "");
	debugf_main("opts.options: %s", opts.options);
	debugf_main("parsed_options: %s", parsed_options);

	if (fuse_opt_add_arg(&fargs, PACKAGE) == -1 ||
	    fuse_opt_add_arg(&fargs, "-s") == -1 ||
	    fuse_opt_add_arg(&fargs, "-o") == -1 ||
	    fuse_opt_add_arg(&fargs, parsed_options) == -1 ||
	    fuse_opt_add_arg(&fargs, opts.mnt_point) == -1) {
		debugf_main("Failed to set FUSE options");
		fuse_opt_free_args(&fargs);
		err = -5;
		goto err_out;
	}

	if (opts.readonly == 0) {
		debugf_main("mounting read-write");
	} else {
		debugf_main("mounting read-only");
	}

	fuse_main(fargs.argc, fargs.argv, &ext2fs_ops, &opts);

err_out:
	fuse_opt_free_args(&fargs);
	free(parsed_options);
	free(opts.options);
	free(opts.device);
	free(opts.volname);
	return err;
}
